# Amyloids “at the border”: deep mutagenesis and random sequence extension reveal an incomplete amyloid-forming motif in Bri2 that turns amyloidogenic upon C-terminal extension

Pipeline to replicate the analysis and figures in: 
<a href="https://www.biorxiv.org/content/10.1101/2023.09.15.557952v1">Amyloids “at the border”: deep mutagenesis and random sequence extension reveal an incomplete amyloid-forming motif in Bri2 that turns amyloidogenic upon C-terminal extension</a>
(Mariano Martín, Benedetta Bolognesi, 2024)


# System requirements
Software and packages to run the pipeline :

- R v3.6.3 (tidyverse, ggplot2, dplyr, reshape2, stringr, readxl, ggpubr, ggrepel, data.table, RColorBrewer, grid, DescTools, GrowthCurver, lsr, esc, ggsignif, pROC)
- DiMSum (https://github.com/lehner-lab/DiMSum; Faure, A.J., Schmiedel, J.M., Baeza-Centurion, P., Lehner B. DiMSum: an error model and pipeline for analyzing deep mutational scanning data and diagnosing common experimental pathologies. Genome Biol 21, 207 (2020))


# Installation
No installation required


# Demo and Instructions for use
- Raw sequencing data and the processed data table required for running the pipeline are deposited in NCBI's Gene Expression Omnibus (GEO) as GSE244612 and GSE270792.
- Needed scripts are located in each folder and properly commented.

